[clang] Correct behavior of `LLVM_UNREACHABLE_OPTIMIZE=OFF` for `Release` builds

MasterAwesome opened the pull request:
# Codegen
### Before
```c++
AArch64SVEPcsAttr *AArch64SVEPcsAttr::CreateImplicit(ASTContext &Ctx, SourceRange Range, Spelling S) {
  AttributeCommonInfo I(Range, NoSemaHandlerAttribute, (
    S == GNU_aarch64_sve_pcs ? AttributeCommonInfo::Form{AttributeCommonInfo::AS_GNU, GNU_aarch64_sve_pcs, false /*IsAlignas*/, false /*IsRegularKeywordAttribute*/} :
    S == CXX11_clang_aarch64_sve_pcs ? AttributeCommonInfo::Form{AttributeCommonInfo::AS_CXX11, CXX11_clang_aarch64_sve_pcs, false /*IsAlignas*/, false /*IsRegularKeywordAttribute*/} :
    S == C23_clang_aarch64_sve_pcs ? AttributeCommonInfo::Form{AttributeCommonInfo::AS_C23, C23_clang_aarch64_sve_pcs, false /*IsAlignas*/, false /*IsRegularKeywordAttribute*/} :
    (llvm_unreachable("Unknown attribute spelling!"),  AttributeCommonInfo::Form{AttributeCommonInfo::AS_GNU, 0, false /*IsAlignas*/, false /*IsRegularKeywordAttribute*/})));
  return CreateImplicit(Ctx, I);
}
```

### After
```c++
AArch64SVEPcsAttr *AArch64SVEPcsAttr::CreateImplicit(ASTContext &Ctx, SourceRange Range, Spelling S) {
  AttributeCommonInfo I(Range, NoSemaHandlerAttribute, [&]() {
    switch (S) {
    case GNU_aarch64_sve_pcs:
      return AttributeCommonInfo::Form{AttributeCommonInfo::AS_GNU, GNU_aarch64_sve_pcs, false /*IsAlignas*/, false /*IsRegularKeywordAttribute*/};
    case CXX11_clang_aarch64_sve_pcs:
      return AttributeCommonInfo::Form{AttributeCommonInfo::AS_CXX11, CXX11_clang_aarch64_sve_pcs, false /*IsAlignas*/, false /*IsRegularKeywordAttribute*/};
    case C23_clang_aarch64_sve_pcs:
      return AttributeCommonInfo::Form{AttributeCommonInfo::AS_C23, C23_clang_aarch64_sve_pcs, false /*IsAlignas*/, false /*IsRegularKeywordAttribute*/};
    default:
      llvm_unreachable("Unknown attribute spelling!");
      return AttributeCommonInfo::Form{AttributeCommonInfo::AS_GNU, 0, false /*IsAlignas*/, false /*IsRegularKeywordAttribute*/};
    }
  }());
  return CreateImplicit(Ctx, I);
}
```

Fixes #68237

MasterAwesome committed:
Correct unreachable hint in release mode

When LLVM_UNREACHABLE_OPTIMIZE is turned off during release mode, a
`do { BUILTIN_TRAP; BUILTIN_UNREACHABLE; } while(0)` is emitted this
causes the ternary operator to not work as expected. Correct this
behavior such that it works in all modes.

Tests:
 * LLVM_UNREACHABLE_OPTIMIZE=[ON|OFF] works in both `Release` and
   `Debug` modes
 * Lambdas on release mode are inlined / similar lambdas are merged.

Signed-off-by: Arvind Mukund <armu30@gmail.com>

MasterAwesome committed:
Remove extra space after `return`

Signed-off-by: Arvind Mukund <armu30@gmail.com>

AaronBallman reviewed (commented):
Oh wow, that bug has been around for quite a while. I like your solution to it!

The code LGTM, no tests required because there isn't really a decent way to test this from lit. However, it would be nice to add a release note to `clang/docs/ReleateNotes.rst` to tell users that the bug was fixed. I'd add the note to the `Bug Fixes in This Version` section in the document.

erichkeane commented:
Agree with Aaron on all points!

MasterAwesome committed:
Add this bug fix to clang's release notes

Signed-off-by: Arvind Mukund <armu30@gmail.com>

MasterAwesome commented:
@erichkeane @AaronBallman, I've added the release notes and rebased the change on `main`.

erichkeane commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -343,6 +343,8 @@ Bug Fixes in This Version
 - Fix a crash when evaluating value-dependent structured binding
   variables at compile time.
   Fixes (`#67690 <https://github.com/llvm/llvm-project/issues/67690>`_)
+- Fixes a regression where ``LLVM_UNREACHABLE_OPTIMIZE=OFF`` cannot be used
```
Can we specify this was a Clang17 regression?

MasterAwesome replied:
Done

AaronBallman reviewed (approved):
LGTM modulo Erich's comment on the release notes.

MasterAwesome committed:
Define the regressed clang version in the release notes

MasterAwesome commented:
@AaronBallman, @erichkeane; whenever ya'll are done with the review, can you merge this in? I don't think I have access. And about the backporting, do we need it for 17.x bugfix releases?

AaronBallman reviewed (approved):
LGTM

AaronBallman commented:
> @erichkeane @AaronBallman, I've added the release notes and rebased the change on `main`. Does this need to be backported to 17.x? 16.x doesn't have this issue

Yeah, I think we should probably backport this to 17.x. We have instructions on how to kick that off here: https://www.llvm.org/docs/GitHub.html#backporting-fixes-to-the-release-branches but I can get the ball rolling by adding the magic comment to the issue.

MasterAwesome commented:
> > @erichkeane @AaronBallman, I've added the release notes and rebased the change on `main`. Does this need to be backported to 17.x? 16.x doesn't have this issue
> 
> Yeah, I think we should probably backport this to 17.x. We have instructions on how to kick that off here: https://www.llvm.org/docs/GitHub.html#backporting-fixes-to-the-release-branches but I can get the ball rolling by adding the magic comment to the issue.

Thanks for doing it! I'll work on cherry-picking this to Rust's fork of llvm :)

AaronBallman commented:
> > > @erichkeane @AaronBallman, I've added the release notes and rebased the change on `main`. Does this need to be backported to 17.x? 16.x doesn't have this issue
> > 
> > 
> > Yeah, I think we should probably backport this to 17.x. We have instructions on how to kick that off here: https://www.llvm.org/docs/GitHub.html#backporting-fixes-to-the-release-branches but I can get the ball rolling by adding the magic comment to the issue.
> 
> Thanks for doing it! I'll work on cherry-picking this to Rust's fork of llvm :)

You're welcome, but it seems there's a problem with trying to cherry-pick this over to 17.x and the pick will be more involved because of merge conflicts with the release notes. Would you mind taking over the pick to 17.x? If not, that's okay! (I'm mostly asking because I'm pretty swamped with reviews and I don't want the backport to fall through the cracks by accident.)

MasterAwesome commented:
> > > > @erichkeane @AaronBallman, I've added the release notes and rebased the change on `main`. Does this need to be backported to 17.x? 16.x doesn't have this issue
> > > 
> > > 
> > > Yeah, I think we should probably backport this to 17.x. We have instructions on how to kick that off here: https://www.llvm.org/docs/GitHub.html#backporting-fixes-to-the-release-branches but I can get the ball rolling by adding the magic comment to the issue.
> > 
> > Thanks for doing it! I'll work on cherry-picking this to Rust's fork of llvm :)
> 
> You're welcome, but it seems there's a problem with trying to cherry-pick this over to 17.x and the pick will be more involved because of merge conflicts with the release notes. Would you mind taking over the pick to 17.x? If not, that's okay! (I'm mostly asking because I'm pretty swamped with reviews and I don't want the backport to fall through the cracks by accident.)

I gotchu! The PR created by llvmbot is at https://github.com/llvm/llvm-project-release-prs/pull/726
